File: live-test/go/generic_super.go
package main

// Super-method fixtures for parameterized and composed interfaces: what a
// method's implicit super set looks like under generics, and how supers
// dedup when embedded interfaces repeat an identical signature.
// Go 1.18+ generic interface — IntStore satisfies Storage[int].
type Storage[T any] interface {
	Get(key string) T
}

type IntStore struct{}

func (s *IntStore) Get(key string) int {
	return 0
}

// Composition: combined interface embedding two interfaces with the same
// method. Tests dedup when a single struct method satisfies multiple
// interfaces with identical signatures.
type Runner1 interface {
	Run() string
}

type Runner2 interface {
	Run() string
}

type Combined interface {
	Runner1
	Runner2
}

type Implementer struct{}

func (i *Implementer) Run() string {
	return "run"
}

File: live-test/rust/src/generic_super.rs
// Generic trait with type parameter — IntStore satisfies Storage<i32>.
pub trait Storage<T> {
    fn get(&self, key: &str) -> T;
}

pub struct IntStore;

// Super on get() resolves through the Storage<i32> instantiation back to the
// generic declaration.
impl Storage<i32> for IntStore {
    fn get(&self, _key: &str) -> i32 {
        0
    }
}
